# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoDecoder.cpp
# Commit: bdfa5574beda
# Full Hash: bdfa5574bedaf23c75fd441c7ebbd41f316a07b8
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2023-06-30 04:06:43
# Regressor Bug: 1831451
# File Overlap Count: 1
# Description:
#   Bug 1831451 - Match display ratio to aspect ratio r=aosmond
#   
#   Depends on D171518
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D179613
# ==============================================================================

diff -r 1ec020b42625 -r bdfa5574beda dom/media/webcodecs/VideoDecoder.cpp
--- a/dom/media/webcodecs/VideoDecoder.cpp	Thu Jun 29 22:33:13 2023 +0000
+++ b/dom/media/webcodecs/VideoDecoder.cpp	Thu Jun 29 22:33:13 2023 +0000
@@ -519,6 +519,38 @@
   return {};
 }
 
+static Result<gfx::IntSize, nsresult> AdjustDisplaySize(
+    const uint32_t aDisplayAspectWidth, const uint32_t aDisplayAspectHeight,
+    const gfx::IntSize& aDisplaySize) {
+  if (aDisplayAspectHeight == 0) {
+    return Err(NS_ERROR_ILLEGAL_VALUE);
+  }
+
+  const double aspectRatio =
+      static_cast<double>(aDisplayAspectWidth) / aDisplayAspectHeight;
+
+  double w = aDisplaySize.width;
+  double h = aDisplaySize.height;
+
+  if (aspectRatio >= w / h) {
+    // Adjust w to match the aspect ratio
+    w = aspectRatio * h;
+  } else {
+    // Adjust h to match the aspect ratio
+    h = w / aspectRatio;
+  }
+
+  w = std::round(w);
+  h = std::round(h);
+  constexpr double MAX = static_cast<double>(
+      std::numeric_limits<decltype(gfx::IntSize::width)>::max());
+  if (w > MAX || h > MAX || w < 1.0 || h < 1.0) {
+    return Err(NS_ERROR_ILLEGAL_VALUE);
+  }
+  return gfx::IntSize(static_cast<decltype(gfx::IntSize::width)>(w),
+                      static_cast<decltype(gfx::IntSize::height)>(h));
+}
+
 // https://w3c.github.io/webcodecs/#create-a-videoframe
 static RefPtr<VideoFrame> CreateVideoFrame(
     nsIGlobalObject* aGlobalObject, const VideoData* aData, int64_t aTimestamp,
@@ -532,7 +564,11 @@
   Maybe<VideoPixelFormat> format = GuessPixelFormat(aData->mImage.get());
   gfx::IntSize displaySize = aData->mDisplay;
   if (mDisplayAspectWidth && mDisplayAspectHeight) {
-    // TODO: Calculate displaySize
+    auto r = AdjustDisplaySize(*mDisplayAspectWidth, *mDisplayAspectHeight,
+                               displaySize);
+    if (r.isOk()) {
+      displaySize = r.unwrap();
+    }
   }
 
   return MakeRefPtr<VideoFrame>(aGlobalObject, aData->mImage, format,
